### PR TITLE
#1152 execute HostConfigModifier at last

### LINKED ISSUE #1152: [Bug]: config portBindings in HostConfigModifier do not work
### Testcontainers version

0.19.0

### Using the latest Testcontainers version?

Yes

### Host OS

macOS

### Host arch

ARM

### Go version

1.19.5

### Docker version

```shell
Client:
 Cloud integration: v1.0.31
 Version:           20.10.23
 API version:       1.41
 Go version:        go1.18.10
 Git commit:        7155243
 Built:             Thu Jan 19 17:35:19 2023
 OS/Arch:           darwin/arm64
 Context:           default
 Experimental:      true

Server: Docker Desktop 4.17.0 (99724)
 Engine:
  Version:          20.10.23
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.10
  Git commit:       6051f14
  Built:            Thu Jan 19 17:31:28 2023
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.18
  GitCommit:        2456e983eb9e37e47538f59ea18f2043c9a73640
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### Docker info

```shell
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.10.3)
  compose: Docker Compose (Docker Inc., v2.15.1)
  dev: Docker Dev Environments (Docker Inc., v0.1.0)
  extension: Manages Docker extensions (Docker Inc., v0.2.18)
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc., 0.6.0)
  scan: Docker Scan (Docker Inc., v0.25.0)
  scout: Command line tool for Docker Scout (Docker Inc., v0.6.0)

Server:
 Containers: 2
  Running: 2
  Paused: 0
  Stopped: 0
 Images: 35
 Server Version: 20.10.23
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 2456e983eb9e37e47538f59ea18f2043c9a73640
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.15.49-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 5
 Total Memory: 7.667GiB
 Name: docker-desktop
 ID: JEWP:W665:LKPJ:LI74:IIQ4:FEOM:FYRH:CXBK:3SVI:N64X:3S42:SWU4
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5000
  127.0.0.0/8
 Live Restore Enabled: false
```

### What happened?

config portBindings in HostConfigModifier do not work

### Relevant log output

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by xmh19936688**
#1152  execute HostConfigModifier at last, or some values will be overwrited

## What does this PR do?

execute HostConfigModifier at last

## Why is it important?

Currently some values such as PortBindings will be overwrited.

## Related issues

- Closes #1152

### CODE REVIEW

**Comment by mdelapenya** [comment]
@xmh19936688 could you extend the existing unit tests to demonstrate the bug? 🙏

File: lifecycle_test.go [hunk 1]
```diff
@@ -300,6 +300,39 @@ func TestPreCreateModifierHook(t *testing.T) {
 			"Networking config's network ID should be retrieved from Docker",
 		)
 	})
+	
+	t.Run("Request contains exposed port modifiers", func(t *testing.T) {
+		// reqWithModifiers {
```
Comment (mdelapenya) [inline lifecycle_test.go:305]: Let's remove this comment, which is used by the docs page to render a code snippet by name
```suggestion
```

File: lifecycle_test.go [hunk 2]
```diff
@@ -300,6 +300,39 @@ func TestPreCreateModifierHook(t *testing.T) {
...
+						{
+							HostIP:   "localhost",
+							HostPort: "8080",
+						},
+					},
+				}
+			},
+			ExposedPorts: []string{"80"},
+		}
+		// }
```
Comment (mdelapenya) [inline lifecycle_test.go:320]: Let's remove this comment, which is used by the docs page to render a code snippet by name
```suggestion
```

**Comment by mdelapenya** [comment]
golangci-lint is complaining that there is a file not go-fmt'ed. Could you check it?

**Comment by mdelapenya** [comment]
Could you double check that the existing tests are passing? I see `TestPreCreateModifierHook` failing on CI too

**Comment by xmh19936688** [comment]
At line 110, it says `Host config's port bindings should be overwritten by the modifier`. But I noticed that `HostIP` and `HostPort` has been modified at line 48 and 49. Why the expectd value is empty string at line 105 and 106? What's the correct behavior?

**Comment by mdelapenya** [comment]
> Why the expectd value is empty string at line 105 and 106?

I think the tests in this case are driven by the implementation/execution of `nat.ParsePortSpecs`, and after debugging that function, the returned binding comes empty for the passed exposed port.

> What's the correct behavior?

What is your idea or use case for it? I remember implementing the lifecycle code as a refactor of the at-that-time existing code, so I tried to replicate it adding unit tests for it.

**Comment by xmh19936688** [comment]
I'd like to specify a host port instead of the random one. So the prerequisite of specifing hostIP and hostPort is it has been exposed, right?

**Comment by mdelapenya** [comment]
> I'd like to specify a host port instead of the random one. So the prerequisite of specifing hostIP and hostPort is it has been exposed, right?

I see your point: because the host config modifier allows to define e.g. custom PortBindings, then they are not applied because the `ParsePortSpecs` call takes place last.

Then I'm OK with changing the order. Do you think we can update the tests to represent the changes in the behaviours?

**Comment by xmh19936688** [comment]
> Do you think we can update the tests to represent the changes in the behaviours?

Sounds good! I think we need some cases to represent how `ExposedPorts` and `HostConfigModifier` effacts.
For example: `ExposedPorts` contains 2379 and 2380, and `HostConfigModifier` binds 2380 to 32380 and 8080 to 38080. What will the result be? (Like this: 2379 to a random hostport, 2380 to 32380, no 8080 ?)

The port in `image.ContainerConfig.ExposedPorts` may also have affect, do they exposed implicitly?

File: lifecycle_test.go (hunk 1 above)
Comment (mdelapenya) [reply lifecycle_test.go:305]: @xmh19936688 I think when we remove this comments, we are ready to go 🚀

Comment (xmh19936688) [reply lifecycle_test.go]: Sorry for didn't see it before. Have done this yet.

**Comment by mdelapenya** [comment]
> The port in `image.ContainerConfig.ExposedPorts` may also have affect, do they exposed implicitly?

We are only exposing those ports automatically if the container request exposes zero ports and the the container does not run in a container network:

```golang
if len(exposedPorts) == 0 && !hostConfig.NetworkMode.IsContainer() {
```

**Comment by xmh19936688** [comment]
> ```go
> if len(exposedPorts) == 0 && !hostConfig.NetworkMode.IsContainer() {
> ```

Apologize for not getting you. Should I do with that `if` like this? 

```go
if len(exposedPorts) == 0 && !hostConfig.NetworkMode.IsContainer() {
    hostConfig.PortBindings = exposedPortMap
} else {
    hostConfig.PortBindings = mergePortBindings(hostConfig.PortBindings, exposedPortMap, req.ExposedPorts)
}
```

**Comment by mdelapenya** [comment]
> ```go
> if len(exposedPorts) == 0 && !hostConfig.NetworkMode.IsContainer() {
>     hostConfig.PortBindings = exposedPortMap
> } else {
>     hostConfig.PortBindings = mergePortBindings(hostConfig.PortBindings, exposedPortMap, req.ExposedPorts)
> }
> ```

I think that will do it: if there are no exposed ports, then keep the behavior of `nat.ParsePortSpecs(exposedPorts)`, else merge with the fix in this PR. We could probably add a comment before the if/else block explaining why we need and not need to merge the port bindings.

wdyt?

**Pull request merged**